sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/ItemsScanBase.scala

```diff
@@ -0,0 +1,262 @@
+  }
+
+  private[this] def toPrettyString(query: SqlQuerySpec) = {
+    //scalastyle:off magic.number
+    val sb = new StringBuilder()
+    //scalastyle:on magic.number
+    sb.append(query.getQueryText)
+    query.getParameters.forEach(
+      (p: SqlParameter) => sb
+        .append(CosmosConstants.SystemProperties.LineSeparator)
+        .append(" > param: ")
+        .append(p.getName)
+        .append(" = ")
+        .append(p.getValue(classOf[Any])))
+
+    sb.toString
+  }
+
+  /**
+   * Returns the actual schema of this data source scan, which may be different from the physical
+   * schema of the underlying storage, as column pruning or other optimizations may happen.
+   */
+  override def readSchema(): StructType = {
+    schema
+  }
+
+  //scalastyle:off method.length
+  override def planInputPartitions(): Array[InputPartition] = {
+    if (plannedInputPartitionsRef.get() == null) {
+      val partitionMetadata = CosmosPartitionPlanner.getFilteredPartitionMetadata(
+        config,
+        clientConfiguration,
+        Some(cosmosClientStateHandles),
+        containerConfig,
+        partitioningConfig,
+        false
+      )
+
+      val calledFrom = s"ItemsScan($description()).planInputPartitions"
+      Loan(
+        List[Option[CosmosClientCacheItem]](
+          Some(CosmosClientCache.apply(
+            clientConfiguration,
+            Some(cosmosClientStateHandles.value.cosmosClientMetadataCaches),
+            calledFrom
+          )),
+          ThroughputControlHelper.getThroughputControlClientCacheItem(
+            config, calledFrom, Some(cosmosClientStateHandles), sparkEnvironmentInfo)
+        ))
+        .to(clientCacheItems => {
+          val container =
+            ThroughputControlHelper.getContainer(
+              config,
+              containerConfig,
+              clientCacheItems(0).get,
+              clientCacheItems(1))
+          SparkUtils.safeOpenConnectionInitCaches(container, log)
+
+          val cosmosInputPartitions = CosmosPartitionPlanner
+            .createInputPartitions(
+              partitioningConfig,
+              container,
+              partitionMetadata,
+              defaultMinPartitionCount,
+              CosmosPartitionPlanner.DefaultPartitionSizeInMB,
+              ReadLimit.allAvailable(),
+              false
+            )
+          plannedInputPartitionsRef.set(cosmosInputPartitions)
+        })
+    }
+
+    if (readManyFiltersMapRef.get() == null) {
+      // there is nothing to prune, return the original planned input partition
+      plannedInputPartitionsRef.get().map(_.asInstanceOf[InputPartition])
+    } else {
+      // only return partitions has matching filter criteria
+      val afterPrunePlannedPartitions = prunePartitions()
+      if (afterPrunePlannedPartitions.size < plannedInputPartitionsRef.get().size) {
+        log.logInfo(s"There are ${plannedInputPartitionsRef.get().size - afterPrunePlannedPartitions.size} partitions got pruned")
+      }
+
```

Comment:
NIT remove 1 LF

Reply:
hmm, can not find the 1 extra lf somehow